### ISSUE TITLE
Integrate qld.gov.au s3 features

### ISSUE THREAD

**Issue by mbocevski** [keitaroinc/ckanext-s3filestore#12]
@ThrawnCA:

> I can't see an option to raise issues on your fork, but you might want to take a look at ours and see if there's any features you want to incorporate.
>
>* Support for using AMI roles instead of hardcoding an access key.
>* Bucket is not checked/created until it's actually used, so command-line invocations (eg paster) have less overhead.
>* Support for substituting the hostname on signed URLs so it's easier to use CloudFront.
>* Better API compatibility with core, so it's easier to interact with other extensions like ckanext-archiver.
>* Specifying mime types on uploads.
>* Support for deleting files from S3, or obtaining metadata.
>* Automatic detection of file size.

Copied from https://github.com/okfn/ckanext-s3filestore/issues/27#issuecomment-733990608.

**Comment by ThrawnCA** [comment]
I notice that you've added a property to specify the addressing type, defaulting to 'auto'. However, I suspect that the code in its current state won't properly work with virtual-style addressing, eg you're providing "endpoint_url" when generating a pre-signed URL. Also, to work as written, your code has to make the "ckanext.s3filestore.host_name" property mandatory.

On the other hand, our code in its current state has to hardcode 'virtual' addressing because it can no longer handle path-style addressing at all. Path-style addressing is deprecated, but undoubtedly still used in a bunch of places.

Probably the ideal option is to have logic in `controller.S3Controller:uploaded_file_redirect` that adjusts the URL format based on the addressing style.

**Comment by mbocevski** [comment]
Sure we can actually make sure that is sent based on config only and use boto to handle format.

**Comment by mbocevski** [comment]
@ThrawnCA we've implemented most of the features, we are left with the delete and detection of file size. Would appreciate if you can check out how things look so far.

**Comment by ThrawnCA** [comment]
The changes I can see on master look pretty good :). I have a small concern about setting the content disposition; doesn't that stop you from viewing PDFs inline in the browser? And there are a few outstanding fixes that we need, such as the ckanext.s3filestore.use_filename property to control the behaviour of old URLs when uploaded resources are replaced with a different filename (if you upload old-filename.pdf and then upload new-filename.csv to the same resource, should /resource/uuid/download/old-filename.pdf point to the old file, or to the new one?)

Also, there's an additional feature that was on develop when I originally raised this, but is now on our master: caching signed URLs in Redis (default 30 minutes).

**Comment by mbocevski** [comment]
Yes the content disposition is implemented so that it's compatible with the default behaviour in CKAN for 'Download Resource', i.e. when a user clicks download then the resource file should be downloaded as a file. One would enable on CKAN the pdf_view plugin and view in order to preview PDFs. Without setting content disposition Chrome will load any text mimetypes within the browser instead of downloading them.

I'll look into the ckanext.s3filestore.use_filename property behaviour. 

Caching signed URLs is a really good idea and we'll cherry-pick that.

**Comment by ThrawnCA** [comment]
Was working on merging some of the Keitaro features into our fork yesterday, and I noticed that there's a script to migrate existing filestore resources to S3, which is something we'd been wanting - but it doesn't work. Fixed in https://github.com/qld-gov-au/ckanext-s3filestore/pull/32/files (updated for Boto3, and added a config parser so it can reuse your INI file instead of hardcoding all the properties it needs).

**Comment by mbocevski** [comment]
@ThrawnCA cool, can you make a PR for that change here on this repo? That would be appreciated.

**Comment by ThrawnCA** [comment]
@mbocevski Pull request #21 covers what we've done thus far.

**Comment by mbocevski** [comment]
@ThrawnCA I think everything is covered now in the extension.

**Comment by ThrawnCA** [comment]
@mbocevski Eh, there's still a bunch of cleanup that could be copied over, like using parameterised logging instead of string preformatting, more robust error handling, and so forth. Plus there's the flag to control the handling of old resource URLs with different filenames; the core uploader will always serve the latest version of the resource, but the default S3 filestore behavior will give you an old resource if you use an old URL.

And it doesn't look like the Redis caching has been copied over?
